fast_lio_sam_qn/include/fast_lio_sam_qn.h

```diff
@@ -0,0 +1,118 @@
+#ifndef FAST_LIO_SAM_QN_MAIN_H
+#define FAST_LIO_SAM_QN_MAIN_H
+
+///// common headers
+#include <time.h>
+#include <math.h>
+#include <cmath>
+#include <chrono> //time check
+#include <vector>
+#include <deque>
+#include <mutex>
+#include <string>
+#include <utility> // pair, make_pair
+#include <tuple>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+///// ROS
+#include <ros/ros.h>
+#include <ros/package.h> // get package_path
+#include <rosbag/bag.h> // save map
+#include <tf/LinearMath/Quaternion.h> // to Quaternion_to_euler
+#include <tf/LinearMath/Matrix3x3.h> // to Quaternion_to_euler
+#include <tf/transform_datatypes.h> // createQuaternionFromRPY
+#include <tf_conversions/tf_eigen.h> // tf <-> eigen
+#include <tf/transform_broadcaster.h> // broadcaster
+#include <std_msgs/String.h>
+#include <geometry_msgs/PoseStamped.h>
+#include <visualization_msgs/Marker.h>
+#include <message_filters/subscriber.h>
+#include <message_filters/time_synchronizer.h>
+#include <message_filters/sync_policies/approximate_time.h>
+///// GTSAM
+#include <gtsam/geometry/Rot3.h>
+#include <gtsam/geometry/Point3.h>
+#include <gtsam/geometry/Pose3.h>
+#include <gtsam/slam/PriorFactor.h>
+#include <gtsam/slam/BetweenFactor.h>
+#include <gtsam/nonlinear/NonlinearFactorGraph.h>
+#include <gtsam/nonlinear/LevenbergMarquardtOptimizer.h>
+#include <gtsam/nonlinear/Values.h>
+#include <gtsam/nonlinear/ISAM2.h>
+///// coded headers
+#include "loop_closure.h"
+#include "pose_pcd.hpp"
+#include "utilities.hpp"
+
+namespace fs = std::filesystem;
+using namespace std::chrono;
+typedef message_filters::sync_policies::ApproximateTime<nav_msgs::Odometry, sensor_msgs::PointCloud2> odom_pcd_sync_pol;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+class FastLioSamQn
+{
+  private:
+    ///// basic params
+    std::string map_frame_;
+    std::string package_path_;
+    std::string seq_name_;
+    ///// shared data - odom and pcd
+    std::mutex realtime_pose_mutex_, keyframes_mutex_;
+    std::mutex graph_mutex_, vis_mutex_;
+    Eigen::Matrix4d last_corrected_pose_ = Eigen::Matrix4d::Identity();
+    Eigen::Matrix4d odom_delta_ = Eigen::Matrix4d::Identity();
+    PosePcd current_frame_;
+    std::vector<PosePcd> keyframes_;
+    int current_keyframe_idx_ = 0;
+    bool is_initialized_ = false;
+    ///// graph and values
+    std::shared_ptr<gtsam::ISAM2> isam_handler_ = nullptr;
```

Comment:
I think all things should be set as `std::unique_ptr` because we don't share those instances with other classes. I leverage the explanations of ChatGPT in Korean for better understanding.

```
unique_ptr은 객체에 대한 독점적 소유권을 제공하며, 리소스가 단일 소유자에 의해 관리되어야 할 때 사용됩니다.
shared_ptr은 객체에 대한 공유 소유권을 제공하며, 여러 소유자가 리소스를 공유해야 할 때 사용됩니다.
```